src/GitVersionCore/ExecuteCore.cs

```diff
@@ -1,12 +1,164 @@
+                var key = string.Format("{0}:{1}:{2}:{3}", gitDir, repo.Head.CanonicalName, repo.Head.Tip.Sha, ticks);
+
+                if (this.getOrAddFromCache != null)
+                {
+                    return this.getOrAddFromCache(key, k =>
+                    {
+                        Logger.WriteInfo("Version not in memory cache. Attempting to load version from disk cache.");
+                        return LoadVersionVariablesFromDiskCache(key, gitDir, targetUrl, dynamicRepositoryLocation, authentication, targetBranch, noFetch, workingDirectory, commitId);
+                    });
+                }
+
+                return LoadVersionVariablesFromDiskCache(key, gitDir, targetUrl, dynamicRepositoryLocation, authentication, targetBranch, noFetch, workingDirectory, commitId);
+            }
+        }
+
+
+        public bool TryGetVersion(string directory, out VersionVariables versionVariables, bool noFetch, Authentication authentication)
+        {
+            try
+            {
+                versionVariables = ExecuteGitVersion(null, null, authentication, null, noFetch, directory, null);
+                return true;
+            }
+            catch (Exception ex)
+            {
+                Logger.WriteWarning("Could not determine assembly version: " + ex);
+                versionVariables = null;
+                return false;
+            }
+        }
+
+
+        static string ResolveCurrentBranch(IBuildServer buildServer, string targetBranch)
+        {
+            if (buildServer == null)
+            {
+                return targetBranch;
+            }
+
+            var currentBranch = buildServer.GetCurrentBranch() ?? targetBranch;
+            Logger.WriteInfo("Branch from build environment: " + currentBranch);
+
+            return currentBranch;
+        }
+
+
+        VersionVariables LoadVersionVariablesFromDiskCache(string key, string gitDir, string targetUrl, string dynamicRepositoryLocation, Authentication authentication, string targetBranch, bool noFetch, string workingDirectory, string commitId)
+        {
+            using (Logger.IndentLog("Loading version variables from disk cache"))
+            {
+                string cacheKey;
+                using (var sha1 = SHA1.Create())
+                {
+                    // Make a shorter key by hashing, to avoid having to long cache filename.
+                    cacheKey = BitConverter.ToString(sha1.ComputeHash(Encoding.UTF8.GetBytes(key))).Replace("-", "");
+                }
+
+                var cacheDir = Path.Combine(gitDir, "gitversion_cache");
+                // If the cacheDir already exists, CreateDirectory just won't do anything (it won't fail). @asbjornu
+                this.fileSystem.CreateDirectory(cacheDir);
+
+                var cacheFileName = string.Concat(Path.Combine(cacheDir, cacheKey), ".yml");
+                VersionVariables vv = null;
+                if (this.fileSystem.Exists(cacheFileName))
```

Comment:
Do we need some sort of cleanup now? Otherwise we will end up with potentially thousands of files in that directory?

Maybe add in a cleanup which deletes everything older than a week?